File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        l_to_r = [1]
        r_to_l = [1]

        for i in range(1,len(nums)):
            l_to_r.append(l_to_r[i-1] * nums[i-1])
        
        for i in range(len(nums)-1,0,-1):
            r_to_l.insert(0, r_to_l[0] * nums[i])
        
        return [l * r for l, r in zip(l_to_r, r_to_l)]

File: Data Structures & Algorithms/jump-game-ii/submission-2.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(),1000000);
        dp[0] = 0;

        for (int i = 0; i < nums.size();i++) {
            if (dp[i] < 1000000) {
                int nextStep = dp[i] + 1;
                for (int k = 1; k < nums[i]+1;k++) {
                    if ((i+k < nums.size()) && (dp[i+k] > nextStep)) {
                        dp[i+k] = nextStep;
                    }
                }
            }
        }

        return dp[nums.size()-1];
    }
};
